KubernetesServer CRUD mode and informers

I'd like to make informers work with the CRUD API. The current behavior seems to break with the reflector code within the informer. For example:

```
    @Test
    void testCrudInformer() throws InterruptedException {
        final KubernetesServer server = new KubernetesServer(true, true);
        server.before();

        KubernetesClient client = server.getClient();
        SharedInformerFactory factory = client.informers();
        SharedIndexInformer<Pod> podInformer = factory.sharedIndexInformerFor(Pod.class, PodList.class, 4000);
        podInformer.addEventHandler(
                new ResourceEventHandler<Pod>() {
                    @Override
                    public void onAdd(Pod obj) {
                    }

                    @Override
                    public void onUpdate(Pod oldObj, Pod newObj) {
                    }

                    @Override
                    public void onDelete(Pod oldObj, boolean deletedFinalStateUnknown) {
                    }
                });
        factory.startAllRegisteredInformers();
        Thread.sleep(5000);
        factory.stopAllRegisteredInformers();
    }
```

Leads to the following NPE in the reflector....

```
Jun 22, 2020 7:15:46 PM okhttp3.mockwebserver.MockWebServer$2 execute
INFO: MockWebServer[58658] starting to accept connections
19:15:46.956 [informer-controller-Pod] INFO  io.fabric8.kubernetes.client.informers.cache.Controller - informer#Controller: ready to run resync and reflector runnable
19:15:46.967 [informer-controller-Pod] INFO  io.fabric8.kubernetes.client.informers.cache.Reflector - Started ReflectorRunnable watch for class io.fabric8.kubernetes.api.model.Pod
Jun 22, 2020 7:15:47 PM okhttp3.mockwebserver.MockWebServer$3 processOneRequest
INFO: MockWebServer[58658] received request: GET /api/v1/namespaces/test/pods HTTP/1.1 and responded: HTTP/1.1 200 OK
19:15:47.397 [informer-controller-Pod] WARN  io.fabric8.kubernetes.client.informers.cache.Controller - Reflector list-watching job exiting because the thread-pool is shutting down
 java.util.concurrent.RejectedExecutionException: Error while starting ReflectorRunnable watch
	at io.fabric8.kubernetes.client.informers.cache.Reflector.listAndWatch(Reflector.java:85) ~[kubernetes-client-4.10.1.jar:?]
	at io.fabric8.kubernetes.client.informers.cache.Controller.run(Controller.java:112) ~[kubernetes-client-4.10.1.jar:?]
	at java.lang.Thread.run(Thread.java:835) [?:?]
Caused by: java.lang.NullPointerException
	at io.fabric8.kubernetes.client.informers.cache.Reflector.reListAndSync(Reflector.java:95) ~[kubernetes-client-4.10.1.jar:?]
	at io.fabric8.kubernetes.client.informers.cache.Reflector.listAndWatch(Reflector.java:80) ~[kubernetes-client-4.10.1.jar:?]
	... 2 more
```

Which in turn happens because the list type (PodList) that the reflector gets back has a null metadata object:
```
  private void reListAndSync() {
    final L list = getList();
    final String latestResourceVersion = list.getMetadata().getResourceVersion(); // NPE: Metadata is null
    ...
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@
 * Fix #2373: Unable to create a Template on OCP3
 * Fix #2316: Cannot load resource from stream without apiVersion
 * Fix #2389: KubernetesServer does not use value from https in crud mode
+* Fix #2306: Make KubernetesServer CRUD mode work with informers
 
 #### Improvements
 * Fix #2331: Fixed documentation for namespaced informer for all custom types implementing `Namespaced` interface

--- a/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesResponseComposer.java
+++ b/kubernetes-server-mock/src/main/java/io/fabric8/kubernetes/client/server/mock/KubernetesResponseComposer.java
@@ -39,7 +39,8 @@ public class KubernetesResponseComposer implements ResponseComposer {
   @Override
   public String compose(Collection<String> collection) {
     return String.format(
-        "{\"apiVersion\":\"v1\",\"kind\":\"List\", \"items\": [%s]}",
+        "{\"apiVersion\":\"v1\",\"kind\":\"List\", \"items\": [%s], " +
+         "\"metadata\": {\"resourceVersion\": \"\", \"selfLink\": \"\"}}",
         join(",", collection));
   }
 }

--- a/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CrudInformerTest.java
+++ b/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/CrudInformerTest.java
@@ -1,0 +1,72 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes.client.mock;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.api.model.PodBuilder;
+import io.fabric8.kubernetes.api.model.PodList;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+import io.fabric8.kubernetes.client.informers.SharedIndexInformer;
+import io.fabric8.kubernetes.client.informers.SharedInformerFactory;
+import io.fabric8.kubernetes.client.server.mock.KubernetesServer;
+import org.junit.Rule;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.migrationsupport.rules.EnableRuleMigrationSupport;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@EnableRuleMigrationSupport
+public class CrudInformerTest {
+  @Rule
+  public KubernetesServer server = new KubernetesServer(true, true);
+
+  // https://github.com/fabric8io/kubernetes-client/issues/2306
+  @Test
+  void testCrudInformer() throws InterruptedException {
+    Pod podToCreate = new PodBuilder().withNewMetadata().withName("pod1").endMetadata().build();
+    KubernetesClient client = server.getClient();
+    SharedInformerFactory factory = client.informers();
+    SharedIndexInformer<Pod> podInformer = factory.sharedIndexInformerFor(Pod.class, PodList.class, 4000);
+    BlockingQueue<Pod> events = new LinkedBlockingQueue<>();
+    podInformer.addEventHandler(
+      new ResourceEventHandler<Pod>() {
+        @Override
+        public void onAdd(Pod obj) {
+          events.add(obj);
+        }
+
+        @Override
+        public void onUpdate(Pod oldObj, Pod newObj) {
+        }
+
+        @Override
+        public void onDelete(Pod oldObj, boolean deletedFinalStateUnknown) {
+        }
+      });
+    factory.startAllRegisteredInformers();
+    client.pods().create(podToCreate);
+    Pod readPod = events.poll(5, TimeUnit.SECONDS);
+    assertNotNull(readPod);
+    assertEquals(readPod.getMetadata().getName(), podToCreate.getMetadata().getName());
+    factory.stopAllRegisteredInformers();
+  }
+}

